### PR TITLE
feat: implement discovery service audit controller

### PR DESCRIPTION
**Pull request by utkuozdemir**
Clean up invalid affiliates from the discovery service if they are no longer part of a cluster from the perspective of Omni.

Introduce a new intermediate resource, `ClusterIdentity`, containing a cluster's unique identifier in the discovery service and all node IDs as they appear ni discovery service as affiliate IDs.

Create the resource, `DiscoveryAuditResult`, to provide a summary on the latest deleted affiliates deleted from the discovery service.

Closes siderolabs/omni#19.

### CODE REVIEW

File: internal/backend/runtime/omni/controllers/omni/discovery_audit.go [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+				logger = logger.With(zap.String("cluster_id", clusterID))
+
+				discoveryAffiliateIDs, err := auditor.listAffiliateIDs(ctx, discoveryClient, clusterID)
+				if err != nil {
+					return err
+				}
+
+				var localNodeIDs []string
+
+				if clusterIdentity.Metadata().Phase() != resource.PhaseTearingDown { // if the cluster is being torn down, we consider all nodes to be invalid
```
Comment (smira) [inline internal/backend/runtime/omni/controllers/omni/discovery_audit.go:76]: can we do potentially something more simple, i.e. drop affiliates for ClusterIdentities which are being torn down?

Comment (utkuozdemir) [reply internal/backend/runtime/omni/controllers/omni/discovery_audit.go]: Done, and reworked the whole thing quite a bit. This was wrong anyway - it needed to be in `FinalizerRemoval`.

(still a draft, will add a couple of more things before it's ready for review).

File: internal/backend/runtime/omni/controllers/omni/discovery_audit.go [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+
+// CreateDiscoveryClientFunc is a function that creates a new DiscoveryClient.
+type CreateDiscoveryClientFunc = func(ctx context.Context) (DiscoveryClient, error)
+
+// DiscoveryAuditController creates the system config patch that contains the maintenance config.
+type DiscoveryAuditController = qtransform.QController[*omni.ClusterIdentity, *omni.DiscoveryAuditResult]
+
+// NewDiscoveryAuditController initializes DiscoveryAuditController.
+func NewDiscoveryAuditController(createDiscoveryClient CreateDiscoveryClientFunc, deleteOlderThan time.Duration) *DiscoveryAuditController {
+	requeueAfter := deleteOlderThan + 1*time.Second
```
Comment (Unix4ever) [inline internal/backend/runtime/omni/controllers/omni/discovery_audit.go:42]:
```suggestion
	requeueAfter := deleteOlderThan + time.Second
```

File: internal/backend/runtime/omni/controllers/omni/discovery_audit.go [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+				logger.Info("found affiliates ready to be deleted")
+
+				deleted, requeue, err := auditor.deleteAffiliates(ctx, discoveryClient, logger, clusterIdentity, cluster, readyToDelete)
+				if err != nil {
+					return err
+				}
+
+				discoveryAuditResult.TypedSpec().Value.DeletedAffiliateIds = deleted
+
+				if requeue {
```
Comment (Unix4ever) [inline internal/backend/runtime/omni/controllers/omni/discovery_audit.go:113]: Why don't we just remove affiliates that do not have corresponding `ClusterMachines` immediately?

Comment (Unix4ever) [inline internal/backend/runtime/omni/controllers/omni/discovery_audit.go:113]: Here I thought about something a bit simpler:
just watch the list of `ClusterMachines` and cleanup the affiliates which don't have a corresponding machine without requeue.

Or even with the used approach: why do we need destroy grace period?

**Comment by utkuozdemir** [comment]
The issue #19 was instead resolved by #149.

This evolved into the https://github.com/siderolabs/omni/pull/150 with some changes.

**Pull request closed without merging**